Flexible widget title input field width

b12f opened the pull request:
When entering a text, the input field now autoresizes with the input text. This gracefully handles overflowing, but
requires browsers to support CSS grid, which all of the browsers in browserslistrc do.

This solution is a bit hacky on the CSS side, but requires almost no javascript to function, and utilizes the browser layouting engine to do width calculations. This is both faster and more accurate, since it accounts for the specific widths of the font being rendered, as well as sub-pixel sizes.

I'm not sure about the coding style, since I was only able to find Ruby standard for that. Feel free to correct that when reviewing.

Closes https://community.openproject.com/projects/openproject/work_packages/32176/activity

b12f committed:
Flexible widget title input field width

When entering a text, the input field now autoresizes with the input text. This gracefully handles overflowing, but
requires browsers to support CSS grid, which all of the browsers in browserslistrc do.

This solution is a bit hacky on the CSS side, but requires almost no javascript to function, and utilizes the browser layouting engine to do width calculations. This is both faster and more accurate, since it accounts for the specific widths of the font being rendered, as well as sub-pixel sizes.

lindenthal commented:
Hi @b12f and @wielinde,

Thanks for working on this.

I was really curious to see the PR in action. Unfortunately, the deployment on pullpreview didn't work. Can you please have a look?

Best
Niels

oliverguenther commented:
Hi @lindenthal @crohr . PullPreview actions are currently not possible for pull requests created outside the OpenProject repository (This PR is coming from a fork).

Cheers
Oliver

wielinde commented:
Hi @b12f,

Thank you for your PR. I just had a quick peek into it and tried it out on my up-to-date Chrome and Firefox on Linux.

Generally speaking it does what the PR asks for.
- [x] The width of the input field adopts to its contents.
- [x] You can click some in the title and the input will start there.

Great!

However, I observed two glitches:

First, when typing the first letter gets cut off frequently:
![first_letter](https://user-images.githubusercontent.com/327272/88186871-191f3f80-cc36-11ea-961d-0f1eb263b405.gif)

Secondly, the CSS classes are used in other places, too such as the titles of work package views in the **Work packages** module. And here 
- the titles are cut off initially and
- the "Save" icon (floppy disk icon) is misplaced.
![saved_query_name_cut_off](https://user-images.githubusercontent.com/327272/88187114-6b606080-cc36-11ea-9781-5b4763fe8267.gif)

Could you please have another go on it?

BTW: I like that it is mainly a CSS solution.

Wieland

wielinde commented:
BTW: I also asked @lindenthal to provide some clarification regarding of the specification of the feature. So please stay tuned. Could you please get yourself an account on community.openproject.com and follow the discussion on the original work package?

b12f committed:
Fixed white-space issue, added outline while editing for editable-toolbar-title

This Fixed an issue where end-of-input whitespace wasn't rendered in the input shim, causing the input field to not
expand. It also adds an outline around the input when focused.

b12f committed:
Fix editable title shim text

The shim text CSS was too specific to the overview page, the new CSS is less specific, taking override rules the same
way the input does.

b12f committed:
Fixed title-container max-width

After removing the overflow: hidden attribute, the title width could be larger than the parent element. This fixes that.

wielinde committed:
Fix unexplicabel text rendering differences.

Adding workaround for unexplicable differences in
text rendering between input and input-shim.
Leading to wrong rendering ellipsis in input-shim.

b12f commented on frontend/src/app/modules/common/editable-toolbar-title/editable-toolbar-title.sass:
```diff
@@ -1,21 +1,57 @@
 
+  &:focus
+    border-color: #999
+    box-shadow: 0 0 0 1px #999
```
This is to get a bigger border on when editing. I'm unsure if you want to go through with this, since it would be the only place where it is different.

We could also make this blue, in which case it would be even more different from the rest of the system. That's fine, as long as you know in which design direction you'd like to be going.

b12f commented on frontend/src/global_styles/layout/_toolbar.sass:
```diff
@@ -204,9 +204,9 @@ $nm-color-success-background: #d8fdd1
-  overflow: hidden
   white-space: nowrap
   margin-bottom: 10px
+  max-width: 100%
```
The overflow: hidden at the top was removed so the fatter "border" on the title input field isn't cut off. However, that made the ellipsis break, which is where the max-width comes in to fix the sizing behaviour.

b12f commented on frontend/src/global_styles/layout/work_packages/_table.sass:
```diff
@@ -54,7 +54,7 @@
     margin-top: 5px
 
-  input[type="text"].toolbar--editable-toolbar
+  .toolbar--editable-toolbar
```
This is a change to make sure the input-shim takes all the style overrides that the input itself has. However, for that to work, the selectors have to target a bit broader than they are currently doing. This decreases selector specificity, and could thus break something, but it looks like everything is fine.

All in all, decreased selector specificity is something that is positive anyway.

b12f commented on frontend/src/global_styles/content/_widget_box.sass:
```diff
@@ -124,7 +124,7 @@ $widget-box--enumeration-width: 20px
   // Ensure style for read-only widget headers
   .editable-toolbar-title--fixed,
-  input[type="text"].toolbar--editable-toolbar
+  .toolbar--editable-toolbar
```
https://github.com/opf/openproject/pull/8516/files#r463654251

b12f commented on frontend/src/global_styles/content/_editable_toolbar.sass:
```diff
@@ -18,7 +18,7 @@
     line-height: 32px
 
-input[type="text"].toolbar--editable-toolbar
+.toolbar--editable-toolbar
```
https://github.com/opf/openproject/pull/8516/files#r463654251

lindenthal commented:
Hi @b12f,

Thanks for your PR. It is already a good improvement when compared with the current implementation.

Since we can not build PR from contributors that are outside this organziation @oliverguenther duplicated your PR in #8540.

It seems as if there is a problem with saving changes. I can not reproduce it properly but there is a problem. Can you please also try to reproduce this?

Thanks
Niels

b12f commented:
Hi Niels,

This issue can be semi-reliably reproduced in the latest stable Docker Version I've got running in my cloud `OpenProject 10.6.4 (PostgreSQL)`, so it appears this PR did not introduce this as a regression.

I've tried to get to the root of the issue, although I haven't been able to fix it. There's some interactions that _could_ trigger this kind of behavior (like changing an `Input()` value from within the component), but I am not familiar enough with Angular to tell you out of the gate what could cause this.

I recommend writing a separate bug report for this.

lindenthal commented:
Hi @b12f,
Thanks for looking into this. I opened a bug report here:
https://community.openproject.com/projects/openproject/work_packages/33998/activity

Best
Niels

oliverguenther commented:
Hi @b12f , the error with saving is unrelated. Somehow the overview grid on the PullPreview environment managed to get itself into an invalid state. To elaborate: The frontend gets a grid with 6 rows, but the coordinates of the widgets contained in it span 7 rows. Once the widget is updated (causing a saving of the entire grid), the error occurs but is not propagated correctly to the user.

In any case, this is looking good. Congrats on your first contribution to OpenProject :tada:
